### ISSUE TITLE
Attributes annotated with a TypeVar with a default lead to Unknown types in pattern matching

### ISSUE THREAD

**Issue by ImogenBits** [microsoft/pyright#7855]
When pattern matching against a class with an attribute that is annotated by a type variable with a default, the attribute's type becomes Unknown in any case blocks. For example:
```py
T = TypeVar("T", bound=int, default=int)

@dataclass
class Thingy(Generic[T]):
    attr: T

def f(obj: Thingy[T]) -> None:
    match obj:
        case Thingy(attr): # here pyright says that the type of attr is `Unknown`
            print(attr)
        case _:
            return
```
The code works fine when replacing the pattern matching with an equivalent if block or when accessing the attribute directly. The bug also still happens when the type variable in the function definition does not have any defaults (but the one in the class does).

I'm using the VSCode extension with pyright version 1.1.360.

**Comment by erictraut** [comment]
This is working as designed, so I don't consider it a bug. However, I think it's a reasonable feature request, so I'm going to change it's label to "enhancement request" and consider it in that context.

**Comment by erictraut** [comment]
Looking at this closer, this is a bug. The behavior depends on whether there is a default value present in the dataclass class definition.

Here's an updated sample:

```python
from dataclasses import dataclass

@dataclass
class Thingy[T: int = int]: # Remove "= int" default to see difference
    attr: T

def f[T: int](obj: Thingy[T]) -> None:
    match obj:
        case Thingy(attr):
            reveal_type(attr)
```

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.362, which I just published.
